src/CargoBuild/CargoBuildTask.cs

```diff
@@ -0,0 +1,437 @@
+                const int RetryCnt = 2;
+
+                path = _cargoPath;
+                args = "fetch";
+
+                Log.LogMessage(MessageImportance.Normal, $"Fetching cargo crates for project in {workingDir}");
+                var exitCode = await ExecuteWithRetriesAsync(path, processArgs: args, workingDir, retryCount: RetryCnt, processRetryArgs: args);
+                Log.LogMessage(MessageImportance.Normal, $"Finished fetching cargo crates for project in  {workingDir}");
+                return exitCode;
+            }
+
+            return authResult;
+        }
+
+        private async Task<ExitCode> ExecuteWithRetriesAsync(string processFileName, string processArgs, string workingDir, int retryCount, string? processRetryArgs = null)
+        {
+            ExitCode exitCode = await ExecuteProcessAsync(processFileName, processArgs, workingDir);
+            const int InitialWaitTimeSec = 3;
+            int retries = 0;
+            while (exitCode != 0 && retries < retryCount)
+            {
+                retries++;
+                int wait = InitialWaitTimeSec * 1000 * retries;
+                Log.LogMessage(MessageImportance.Normal, $"Process failed with exit code: {exitCode}. Retry #{retries}: {processFileName}. Waiting {wait / 600} seconds before retrying.");
+
+                await Tasks.Task.Delay(wait);
+                exitCode = await ExecuteProcessAsync(processFileName, processRetryArgs ?? processArgs, workingDir);
+            }
+
+            return exitCode;
+        }
+
+        private async Task<ExitCode> DoRegistryAuthAsync(string workingDir)
+        {
+            return await ExecuteWithRetriesAsync(_cargoPath, $"login --registry {RegistryFeedName}", workingDir, retryCount: 2);
+        }
+
+        private async Task<ExitCode> ExecuteProcessAsync(string fileName, string args, string workingDir, Dictionary<string, string>? envars = null)
+        {
+            try
+            {
+                var processTask = new Task<int>(() =>
+                {
+                    var info = new ProcessStartInfo
+                    {
+                        RedirectStandardError = true,
+                        RedirectStandardOutput = true,
+                        UseShellExecute = false,
+                        FileName = fileName,
+                        Arguments = args,
+                        WorkingDirectory = workingDir,
+                    };
+                    if (envars != null && envars.Count > 0)
+                    {
+                        foreach (var envar in envars)
+                        {
+                            if (!info.EnvironmentVariables.ContainsKey(envar.Key))
+                            {
+                                info.EnvironmentVariables.Add(envar.Key, envar.Value);
+                            }
+                        }
+                    }
+
+                    var process = Process.Start(info);
+                    int maxWait = 1000 * 60 * 15;
```

Comment:
Do we need a timeout? Other SDKs, eg for .NET, do not enforce a timeout on the compiler